command/ca/certificate.go

```diff
@@ -443,3 +486,408 @@ func parseTimeDuration(ctx *cli.Context) (notBefore api.TimeDuration, notAfter a
+		}
+		if ro.Status == "ready" {
+			isReady = true
+			ui.Printf(" done!\n")
+			break
+		}
+	}
+	if !isReady {
+		ui.Printf(" Error!\n\n")
+		return nil, errors.Errorf("Unable to validate order: %+v", ro)
+	}
+
+	ui.Printf("Finalizing Order .")
+	if err = ac.FinalizeOrder(o.Finalize, csr); err != nil {
+		return nil, errors.Wrapf(err, "error finalizing order")
+	}
+
+	for i := 9; i >= 0; i-- {
+		time.Sleep(1 * time.Second)
+		ui.Printf(".")
+		fo, err = ac.GetOrder(o.ID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error retrieving order %s", o.ID)
+		}
+		if fo.Status == "valid" {
+			isValid = true
+			ui.Printf(" done!\n")
+			break
+		}
+	}
+	if !isValid {
+		ui.Printf(" Error!\n\n")
+		return nil, errors.Errorf("Unable to finalize order: %+v", fo)
+	}
+
+	return fo, nil
+}
+
+func acmeFlow(ctx *cli.Context) error {
+	args := ctx.Args()
+	subject := args.Get(0)
+	crtFile, keyFile := args.Get(1), args.Get(2)
+
+	sans := ctx.StringSlice("san")
+
+	if len(sans) == 0 {
+		sans = []string{subject}
+	}
+	dnsNames, ips := splitSANs(sans)
+	if len(ips) > 0 {
+		return errors.New("IP Address SANs are not supported for ACME flow")
+	}
+
+	caURL := ctx.String("ca-url")
+	if len(caURL) == 0 {
+		return errs.RequiredFlag(ctx, "ca-url")
+	}
+	if !(strings.HasSuffix(caURL, "directory") || strings.HasSuffix(caURL, "dir")) {
+		caURL = caURL + "/acme/directory"
+	}
+
+	var idents []acme.Identifier
+	for _, dns := range dnsNames {
+		idents = append(idents, acme.Identifier{
+			Type:  "dns",
+			Value: dns,
+		})
+	}
+
+	var (
+		err          error
+		orderPayload []byte
+		clientOps    []ca.ClientOption
+	)
+	if strings.Contains(caURL, "letsencrypt") {
+		// LetsEncrypt does not support NotBefore and NotAfter attributes in orders.
+		if ctx.IsSet("not-before") || ctx.IsSet("not-after") {
+			return errors.New("LetsEncrypt public CA does not support NotBefore/NotAfter " +
+				"attributes for certificates. Instead, each certificate has a default lifetime of 3 months.")
+		}
+		// Use default transport for public CAs
+		clientOps = append(clientOps, ca.WithTransport(http.DefaultTransport))
+		// LetsEncrypt requires that the Common Name of the Certificate also be
+		// represented as a DNSName in the SAN extension, and therefore must be
+		// authorized as part of the ACME order.
+		hasSubject := false
+		for _, n := range idents {
+			if n.Value == subject {
+				hasSubject = true
+			}
+		}
+		if !hasSubject {
+			dnsNames = append(dnsNames, subject)
+			idents = append(idents, acme.Identifier{
+				Type:  "dns",
+				Value: subject,
+			})
+		}
+		orderPayload, err = json.Marshal(struct {
+			Identifiers []acme.Identifier
+		}{Identifiers: idents})
+		if err != nil {
+			return errors.Wrap(err, "error marshaling new letsencrypt order request")
+		}
+	} else {
+		// If the CA is not public then a root file is required.
+		root := ctx.String("root")
+		if len(root) == 0 {
+			return errs.RequiredFlag(ctx, "root")
+		}
```

Comment:
Do you mean the system defaults, or the default step root CA (as defined in defaults.json)?

Reply:
Ours, if it exists, some of our code already checks for the known path where step root cert should be, i.e:
https://github.com/smallstep/cli/blob/master/command/ca/certificate.go#L249-L254

Reply:
So this is for a case when root is not defined anywhere but we still try and check for it in the `$(step path)`, just in case? Just want to make sure I'm clear on the use case.

Reply:
yes, when `defaults.json` does not exist or root is not defined there, but they have `~/.step/certs/root_ca.crt`, to be consistent with other behavior of `step ca certificate`